Firestore/core/src/local/leveldb_key.h

```diff
@@ -1026,46 +1094,73 @@ class LevelDbDocumentOverlayLargestBatchIdIndexKey {
-  }
+  static std::string KeyPrefix();
 
-  // Overload `ToLevelDbDocumentOverlayKey()` to avoid copies if invoked on
-  // an rvalue reference.
-  LevelDbDocumentOverlayKey ToLevelDbDocumentOverlayKey() && {
-    return LevelDbDocumentOverlayKey(std::move(user_id_),
-                                     std::move(document_key_),
-                                     std::move(largest_batch_id_));
-  }
+  /**
+   * Creates a key prefix that points just before the first key for the given
+   * user_id.
+   */
+  static std::string KeyPrefix(absl::string_view user_id);
 
-  /** The user ID, as encoded in the key. */
-  const std::string& user_id() const {
-    return user_id_;
-  }
+  /**
+   * Creates a key prefix that points just before the first key for the given
+   * user_id, collection.
+   */
+  static std::string KeyPrefix(absl::string_view user_id,
+                               const model::ResourcePath& collection);
 
-  /** The largest_batch_id_, as encoded in the key. */
-  model::BatchId largest_batch_id() const {
-    return largest_batch_id_;
+  /**
+   * Creates a key prefix that points just before the first key for the given
+   * user_id, collection, and largest_batch_id.
+   */
+  static std::string KeyPrefix(absl::string_view user_id,
+                               const model::ResourcePath& collection,
+                               model::BatchId largest_batch_id);
+
+  /**
+   * Creates a complete key that points to a specific user_id, collection, and
+   * largest_batch_id for a given key in the document_overlays table.
+   */
+  static std::string Key(absl::string_view user_id,
+                         const model::ResourcePath& collection,
+                         model::BatchId largest_batch_id,
+                         const model::DocumentKey& document_key);
+
+  /**
+   * Creates a complete key that points to a specific key in the overlays table.
+   */
+  static std::string Key(const LevelDbDocumentOverlayKey& key) {
+    return Key(key.user_id(), key.document_key().path().PopLast(),
+               key.largest_batch_id(), key.document_key());
   }
 
   /**
-   * The key in the document_overlays table to which this index entry points,
-   * as encoded in the key.
+   * Decodes the given complete key, storing the decoded values in this
+   * instance.
+   *
+   * @return true if the key successfully decoded, false otherwise. If false is
+   * returned, this instance is in an undefined state until the next call to
+   * `Decode()`.
    */
-  const model::DocumentKey& document_key() const {
-    return document_key_;
+  ABSL_MUST_USE_RESULT
+  bool Decode(absl::string_view key);
+
+  /** The collection_, as encoded in the key. */
+  const model::ResourcePath& collection() const {
+    return collection_;
   }
 
  private:
-  std::string user_id_;
-  model::BatchId largest_batch_id_ = -1;
-  model::DocumentKey document_key_;
+  model::ResourcePath collection_;
```

Comment:
FWIW, this should be part of the document_key_ in the superclass.

Reply:
Done. I've removed the `collection_` member variable.